File: libs/core/src/binder/terminal_binder.cpp
/**
 * @file terminal_binder.cpp
 * @author Francesco Corso
 * @date 8/17/2026
 * @brief
 */
module prism.core:binder.terminal_binder.impl;

import :binder.terminal_binder;

namespace prism
{

    TerminalBinder::TerminalBinder(const Compilation &compilation) : Binder(compilation)
    {
    }

    Optional<const Binder &> TerminalBinder::get_binder(const SyntaxNode &node) const
    {
        return std::nullopt;
    }

    Optional<const Symbol &> TerminalBinder::containing_symbol() const
    {
        return std::nullopt;
    }

    Optional<const SyntaxNode &> TerminalBinder::scope_designator() const
    {
        return std::nullopt;
    }

    VariablesSpan TerminalBinder::get_declared_local_variables_for_scope(const SyntaxNode &) const
    {
        return {};
    }

    LookupResult TerminalBinder::lookup_local(Name) const
    {
        return LookupResult{};
    }
} // namespace prism

File: src/Prism.Core/Binding/TerminalBinder.cs
using Prism.Core.Compiling;
using Prism.Core.Semantic;
using Prism.Core.Symbols;
using Prism.Core.Syntax;

namespace Prism.Core.Binding;

internal sealed class TerminalBinder(Compilation compilation) : Binder(compilation)
{
    protected override Symbol? ContainingSymbol => null;

    protected override SyntaxNode? ScopeDesignator => null;

    public override Binder? GetBinder(SyntaxNode node) => null;

    protected override void EnsureLocals()
    {
        // Do nothing
    }

    protected override LookupResult LookupLocal(
        string name,
        LookupOptions options,
        BindingContext context
    )
    {
        return LookupResult.NotFound();
    }

    public override LabelSymbol? LookupLoopLabel(string name, BindingContext context) => null;
}
